### PR TITLE
Add ability to create a DSS client from FM client

### PR DESCRIPTION
**Pull request by amandineslx**
[sc-95097]

Be able to create a DSS client using the node ID and an API key authorized on this node in order to perform actions on the DSS instance.

This PR is paired with [dip PR](https://github.com/dataiku/dip/pull/17499).

### CODE REVIEW

**Comment by ericcournarie** [comment]
Tested with the following workflow
```

import dataiku
import dataikuapi

api = dataiku.api_client ()
cluster = api.get_cluster ("EricK8S")
print (cluster.get_status().get_raw())

// use fmadmin create-personal-api-key to create key
aws = dataikuapi.FMClientAWS ("https://<FM host>", "<Key ID>", "<Key secret>")
aws._session.verify = None

// instance id is in the URL, ( not written anywhere ? )
awscli = aws.get_instance_client ("li-pocLR7outIDb", "<key secret of target DSS>")
awscli._session.verify = None

cluster = awscli.get_cluster ("EricK8S")
print (cluster.get_status().get_raw())


```

**Comment by ericcournarie** [comment]
as a nitpick, I would issue a warn log if govern is asked for, as we may forget the comment .. something like ' govern is not yet supported... '
did not check what it does now, but as it seems, will not be very effective ?

File: dataikuapi/fmclient.py [hunk 1]
```diff
@@ -210,6 +213,27 @@ def list_instance_images(self):
...
+        :rtype: :class:`dataikuapi.dssclient.DSSClient`
+        """
+        instance = self.get_instance(instance_id)
+        instance_status = instance.get_status()
+        public_url = instance_status.get("publicURL")
+
+        # TODO waiting for PR merge to be uncommented
+        # https://github.com/dataiku/dataiku-api-client-python/pull/245
+        # if instance.instance_data.get("nodeType") == "govern":
+            # return GovernClient(public_url, api_key)
```
Comment (ericcournarie) [inline dataikuapi/fmclient.py:233]: a warning message will be welcome, we may forget the comment ...

**Review by ericcournarie — APPROVED** [review]
LGTM

File: dataikuapi/fmclient.py (hunk 1 above)
Comment (amandineslx) [reply dataikuapi/fmclient.py:234]: I prefer keeping the comment as the PR is already ongoing
we will probably have to coordinate with them:
* if we merge first, they uncomment the code in their PR
* if they merge first, we remove the exception raising and implement directly the `GovernClient` correctly

**Comment by ericcournarie** [comment]
script update
```

api = dataiku.api_client ()
prj = api.get_project ("DKU_TSHIRTS")
print (prj.get_summary()["projectKey"])


aws = dataikuapi.FMClientAWS ("https://<IP>", "<key>", "<secret>")
aws._session.verify = None

awscli = aws.get_instance("li-FD0NxffN2AKz").get_client ()
awscli._session.verify = None

prj = awscli.get_project ("DKU_TSHIRTS")
print (prj.get_summary()["projectKey"])
 
```

File: dataikuapi/fm/instances.py [hunk 2]
```diff
@@ -183,6 +186,28 @@ def __init__(self, client, instance_data):
...
+        api_key = self.instance_data.get("adminAPIKey")
+
+        if self.instance_data.get("dssNodeType") == "govern":
+            # TODO waiting for PR merge to be uncommented
+            # https://github.com/dataiku/dataiku-api-client-python/pull/245
+            raise Exception("DSS client for Govern nodes is not available")
+            # return GovernClient(public_url, api_key)
+
+        if not public_url:
+            raise ValueError("No public URL available for node %s. This node may not be provisioned yet" % instance_id)
```
Comment (thtrunck) [inline dataikuapi/fm/instances.py:207]: Does the instance_id actually exist? Shouldn't it be `self.id`?

**Review by thtrunck — APPROVED** [review]
LGTM
Tested it on a patched FM instance.
But with the move of the client into dip we will need to redo this PR in DIP after the merge

**Comment by amandineslx** [comment]
closed on behalf of [DIP PR](https://github.com/dataiku/dip/pull/18045) following Python API client import in DIP repository

**Pull request closed without merging**